### PR TITLE
Add USD UNI-V2 LP token price identifiers

### PR DESCRIPTION
**Pull request by dev-1-lp-dollar**
This is a proposal to add USD-UNI-V2-WBTC-ETH as a supported price identifier.

The price identifier will allow LP Dollar to offer fixed borrowing costs to UNI-V2-WBTC-ETH liquidity providers, and serve as a tool for the UMA ecosystem to create synthetics with UNI-V2-WBTC-ETH LP positions as collateral.

### CODE REVIEW

**Comment by smb2796** [comment]
Seems like all of the price identifiers that you just submitted could be housed in one price id umip. Same for all of the collateral currencies.

**Comment by dev-1-lp-dollar** [comment]
The trouble is that each pair has different underlying oracle requirements (for example, getting the WBTC vs the UMA price). But I can combine all of them if you would like.

**Comment by smb2796** [comment]
>The trouble is that each pair has different underlying oracle requirements (for example, getting the WBTC vs the UMA price). But I can combine all of them if you would like.

That would be best, and then just specifying the different requirements for each within the same UMIP

**Comment by dev-1-lp-dollar** [comment]
> > The trouble is that each pair has different underlying oracle requirements (for example, getting the WBTC vs the UMA price). But I can combine all of them if you would like.
> 
> That would be best, and then just specifying the different requirements for each within the same UMIP

Combined them here and for the collateral UMIP.

**Comment by smb2796** [comment]
Awesome, thanks for doing this. Because the UMIP submissions missed the weekly submission deadline, we unfortunately need to move these to next week's review cycle. The great thing about this is that we are already building out a price feed that should work for these! The price feed should be finished by next week, so we will be setup well to vote in the following cycle.

**Comment by dev-1-lp-dollar** [comment]
> Awesome, thanks for doing this. Because the UMIP submissions missed the weekly submission deadline, we unfortunately need to move these to next week's review cycle. The great thing about this is that we are already building out a price feed that should work for these! The price feed should be finished by next week, so we will be setup well to vote in the following cycle.

Yeah, no problem, thanks for the feedback. Regarding timeline, meaning it will be discussed on Monday and voted on next Friday? Not too familiar with the cycles.

**Comment by dev-1-lp-dollar** [comment]
According to the calendar, as far as I can tell, it looks like the PR was opened on Wednesday before the deadline to meet the review on Monday.

**Comment by dev-1-lp-dollar** [comment]
Yeah it looks like I posted in Discord community and Discourse in time but took my time PRing it in here. :/

File: UMIPs/umip-usd-uni-v2-lp.md [hunk 1]
```diff
@@ -0,0 +1,766 @@
...
+  
+    = (1 / 520.3440264478901) * 1e18 
+  
+    = 1921805477092654  
+    ```
+
+
+## Technical Specifications
+
+Price Identifier Name: USD-UNI-V2-WBTC-ETH
```
Comment (Tommy1231232) [inline UMIPs/umip-usd-uni-v2-lp.md:727]: For all the names lets update to USD/UNI-V2-WBTC-ETH. (this makes it easier to distinguish the base and quote currency)

File: UMIPs/umip-usd-uni-v2-lp.md [hunk 2]
```diff
@@ -0,0 +1,766 @@
...
+| Created    | February 09, 2021           |
+| Link to Discourse | [Discourse](https://discourse.umaproject.org/t/add-usd-uni-v2-wbtc-eth-as-a-price-identifier/148)                                 |
+
+## Summary
+
+The DVM should support price requests for the following Uniswap V2 LP tokens: WBTC-ETH, USDC-ETH, UNI-ETH, UMA-ETH.
+
+## Motivation
+
+The DVM currently does not support Uniswap V2 LP tokens.
```
[the remark this answers is not in the document]
Comment (dev-2-lp-dollar) [reply UMIPs/umip-usd-uni-v2-lp.md:16]: Correct from my understanding. V1 or V3 would be separate liquidity pools from what this price ID is expected to reference for V2. 

Because they're separate liquidity pools, they have a separate exchange rate, and therefore can't be treated equal. It's reasonable to expect the prices across these different pools (of the same underlying assets) to be quite close due to the nature of arbitrage. However, I wouldn't expect them to be equal. 

Let me know if this line of thought makes sense, or if you had a different POV please!

File: UMIPs/umip-usd-uni-v2-lp.md (hunk 1 above)
Comment (dev-2-lp-dollar) [reply UMIPs/umip-usd-uni-v2-lp.md:727]: @Tommy1231232 good idea :+1: , will do.

File: UMIPs/umip-usd-uni-v2-lp.md [hunk 3]
```diff
@@ -0,0 +1,766 @@
...
+
+By enabling LP token price identifiers, UMA will open the door for the creation of synths with CFMM (constant function market maker) LP positions.
+
+The LP Dollar team will use the price identifiers to enable fixed borrowing costs for Uniswap V2 liquidity providers.
+
+## Data Sources & Price Feed Implementation
```
Comment (Reinis-FRP) [inline UMIPs/umip-usd-uni-v2-lp.md:22]: Can you split this out to `MARKETS & DATA SOURCES`, `PRICE FEED IMPLEMENTATION` and `IMPLEMENTATION` sections?
Do you have a link to reference implementation script that combines all steps described in this section?

File: UMIPs/umip-usd-uni-v2-lp.md [hunk 4]
```diff
@@ -0,0 +1,766 @@
...
+Rounding: Round to nearest 18 decimal places (nineteenth decimal place digit >= 5 rounds up and < 5 rounds down)  
+
+Price Identifier Name: USD-UNI-V2-UMA-ETH  
+Base Currency: USD  
+Quote currency: UNI-V2-UMA-ETH  
+Intended Collateral Currency: UNI-V2-UMA-ETH  
+Collateral Decimals: 18  
+Rounding: Round to nearest 18 decimal places (nineteenth decimal place digit >= 5 rounds up and < 5 rounds down)
+
+## Rationale
```
Comment (Reinis-FRP) [inline UMIPs/umip-usd-uni-v2-lp.md:755]: Can you elaborate the rationale section on the design motivation (e.g. reasoning for selecting particular markets, price robustness, potential for price manipulation)?

File: UMIPs/umip-usd-uni-v2-lp.md [hunk 5]
```diff
@@ -0,0 +1,766 @@
...
+
+## Security Considerations
+Adding these new price identifiers should not effect the security to the DVM or priceless financial contract users. However, anyone deploying a new priceless token contract referencing these identifiers should take care to parameterize the contract appropriately to avoid the loss of funds for synthetic token holders. Additionally, the contract deployer should ensure that there is a network of liquidators and disputers ready to perform the services necessary to keep the contract solvent.
+
+The Uniswap smart contracts are among the highest quality in decentralized finance, safely locking up and handling billions of dollars in different assets.
+
+Uniswap as a price feed is online 24/7 with no downtime as it's a protocol hosted on the Ethereum blockchain. Therefore, there will be no forced existence of time gaps in price data - it will always be available.
+
+These pairs are susceptible to price volatility, and have moved in the ranges of 10-30% within 24 hours, with the more extreme side of the range highly unlikely, but still possible. A high collateralization requirement (120%+) should be set to mitigate this potential volatility risk.
```
Comment (mhairimca) [inline UMIPs/umip-usd-uni-v2-lp.md:766]: Are there any risks to manipulation of the balance of the assets within a Uniswap pool using flashloans?

File: UMIPs/umip-usd-uni-v2-lp.md [hunk 6]
```diff
@@ -0,0 +1,766 @@
...
+
+    Base Currency: UNI  
+    Quote Currency: USD  
+
+    Exchanges: Coinbase Pro (UNI:USD), Kraken (UNI:USD), Bitfinex (UNI:USD)
+    Input Processing: None. Human intervention in extreme circumstances where the result differs from broad market consensus.  
+    Price Steps: 0.01 (2 decimals in more general trading format)  
+    Rounding: Closest, 0.5 up  
+    Pricing Interval: 60 seconds  
+    Dispute timestamp rounding: down
```
Comment (akhanubis) [inline UMIPs/umip-usd-uni-v2-lp.md:503]: Can you add Result Processing: Median to make it clearer? Also applies to UMA:USD.

**Review by akhanubis — COMMENTED** [review]
There are some extra `)` in the price calculation examples

File: UMIPs/umip-usd-uni-v2-lp.md (hunk 6 above)
Comment (mhairimca) [reply UMIPs/umip-usd-uni-v2-lp.md:503]: Should the decimals here be 18 to match the collateral currency?
@smb2796  - could you clarify this as I'm not sure.

File: UMIPs/umip-usd-uni-v2-lp.md (hunk 2 above)
[the remark this answers is not in the document]
Comment (pemulis) [reply UMIPs/umip-usd-uni-v2-lp.md:16]: > Correct from my understanding. V1 or V3 would be separate liquidity pools from what this price ID is expected to reference for V2.
> 
> Because they're separate liquidity pools, they have a separate exchange rate, and therefore can't be treated equal. It's reasonable to expect the prices across these different pools (of the same underlying assets) to be quite close due to the nature of arbitrage. However, I wouldn't expect them to be equal.
> 
> Let me know if this line of thought makes sense, or if you had a different POV please!

Agreed!

File: UMIPs/umip-usd-uni-v2-lp.md [hunk 7]
```diff
@@ -0,0 +1,766 @@
...
+    ```
+
+4) Fourth, use the UNI-V2-USDC-ETH total supply of LP tokens and USD value of reserves to calculate the USD value of each LP token (UNI-V2-USDC-ETH:USD). USDC is treated as equal to USD.
+
+    Example calculation of UNI-V2-USDC-ETH:USD:
+
+    ```
+    UNI-V2-USDC-ETH:USD =
+  
+    (USD value of USDC reserves + USD value of WETH reserves) / UNI-V2-USDC-ETH total supply of LP tokens)
```
[the remark this answers is not in the document]
Comment (pemulis) [reply UMIPs/umip-usd-uni-v2-lp.md:342]: I think it makes sense to include that additional step, since USDC is almost always slightly off-peg in one direction or the other, and the effect is amplified by pooling a bunch of USDC together.

File: UMIPs/umip-usd-uni-v2-lp.md [hunk 8]
```diff
@@ -0,0 +1,766 @@
...
+  
+    = (1 / 520.3440264478901) * 1e18 
+  
+    = 1921805477092654  
+    ```
+
+
+## Technical Specifications
+
+Price Identifier Name: USD/UNI-V2-WBTC-ETH
```
Comment (pemulis) [inline UMIPs/umip-usd-uni-v2-lp.md:727]: @mrice32 would the `/` in the price identifier cause any problems with our bot clients, or would it be escaped automatically? Cf. the expression price feed.

File: UMIPs/umip-usd-uni-v2-lp.md (hunk 5 above)
Comment (pemulis) [reply UMIPs/umip-usd-uni-v2-lp.md:766]: I think that the query will return the balance of the assets at the beginning of the specified block, before any transactions, but I could be wrong.

File: UMIPs/umip-usd-uni-v2-lp.md [hunk 9]
```diff
@@ -0,0 +1,766 @@
...
+
+The DVM should support price requests for the following Uniswap V2 LP tokens: WBTC-ETH, USDC-ETH, UNI-ETH, UMA-ETH.
+
+## Motivation
+
+The DVM currently does not support Uniswap V2 LP tokens.
+
+By enabling LP token price identifiers, UMA will open the door for the creation of synths with CFMM (constant function market maker) LP positions.
+
+The LP Dollar team will use the price identifiers to enable fixed borrowing costs for Uniswap V2 liquidity providers.
```
[the remark this answers is not in the document]
Comment (dev-1-lp-dollar) [reply UMIPs/umip-usd-uni-v2-lp.md:20]: It is similar to the yield dollar ETH & renBTC fixed borrowing that the UMA team has been supporting. I can elaborate, but it was more just mean to be a brief mention of what is possible with the price identifier.

File: UMIPs/umip-usd-uni-v2-lp.md [hunk 10]
```diff
@@ -0,0 +1,766 @@
...
+
```
[the remark this answers is not in the document]
Comment (dev-1-lp-dollar) [reply UMIPs/umip-usd-uni-v2-lp.md:21]: Okay I can elaborate on what we'll be using it for

File: UMIPs/umip-usd-uni-v2-lp.md (hunk 3 above)
Comment (dev-1-lp-dollar) [reply UMIPs/umip-usd-uni-v2-lp.md:22]: We haven't released a reference implementation of the script but surely can. 

Note, we wouldn't want to conflict with anything the UMA team is putting out, however, and in the discord there was mention of UMA building a price feed for this UMIP. Can someone on the UMA team comment here?

File: UMIPs/umip-usd-uni-v2-lp.md [hunk 11]
```diff
@@ -0,0 +1,766 @@
...
+     ```
+
+3) Third, the ETHUSD price must be queried and used to calculate the USD value of the WETH reserves. 
+
+    The methodology to query the price of ETHUSD is similar to [UMIP-6](https://github.com/UMAprotocol/UMIPs/blob/master/UMIPs/umip-6.md), with a different exchange list and decimal precision.  
+
+    Base Currency: ETH  
+    Quote Currency: USD  
+
+    Exchanges: Coinbase Pro (ETH:USD), Kraken (ETH:USD), Bitfinex (ETH:USD), Bitstamp (ETH:USD)
```
[the remark this answers is not in the document]
Comment (dev-1-lp-dollar) [reply UMIPs/umip-usd-uni-v2-lp.md:134]: We can recommend endpoints for sure. I'm not sure what you mean about an api key though.

File: UMIPs/umip-usd-uni-v2-lp.md [hunk 12]
```diff
@@ -0,0 +1,766 @@
...
+
+    Base Currency: ETH  
+    Quote Currency: USD  
+
+    Exchanges: Coinbase Pro (ETH:USD), Kraken (ETH:USD), Bitfinex (ETH:USD), Bitstamp (ETH:USD)  
+    Input Processing: None. Human intervention in extreme circumstances where the result differs from broad market consensus.  
+    Price Steps: 0.01 (2 decimals in more general trading format)  
+    Rounding: Closest, 0.5 up  
+    Pricing Interval: 60 seconds  
+    Dispute timestamp rounding: down
```
[the remark this answers is not in the document]
Comment (dev-1-lp-dollar) [reply UMIPs/umip-usd-uni-v2-lp.md:139]: good question - i think relative to the requested price feed timestamp makes the most sense to me. will add the clarification if that sounds good

File: UMIPs/umip-usd-uni-v2-lp.md [hunk 13]
```diff
@@ -0,0 +1,766 @@
...
+    
+    = 97499.896966146357068372 * 1716.12 
+    
+    = 167321523.18154308
+    ```
+    
+4) Fourth, the BTC:USD price must be queried and used to calculate the USD value of the WBTC reserves. 
+
+    The methodology to query the price of BTC:USD is similar to [UMIP-7](https://github.com/UMAprotocol/UMIPs/blob/master/UMIPs/umip-7.md), with a different exchange list and decimal precision.  
+
```
[the remark this answers is not in the document]
Comment (dev-1-lp-dollar) [reply UMIPs/umip-usd-uni-v2-lp.md:156]: For BTC:USD, we think two decimals is appropriate given that most endpoints we found only returned that many anyway. For the BTC:USD price, this will not effect the calculation much.

Decimals are important in the steps that follow getting the BTC:USD price, so we can be more specific there about the decimals there.

File: UMIPs/umip-usd-uni-v2-lp.md (hunk 7 above)
[the remark this answers is not in the document]
Comment (dev-1-lp-dollar) [reply UMIPs/umip-usd-uni-v2-lp.md:342]: I think it is fair to treat USDC as equal to 1 USD. I suspect users would expect it to be treated as such, and given free exchange of USDC for USD on Coinbase Pro I think it is reasonable given market structure. 

If it trades off peg anywhere, I would suspect that to be due to trading fees on certain exchanges. But would be interested to see anywhere it is trading off peg.

File: UMIPs/umip-usd-uni-v2-lp.md (hunk 6 above)
Comment (dev-1-lp-dollar) [reply UMIPs/umip-usd-uni-v2-lp.md:503]: Sure thing!

File: UMIPs/umip-usd-uni-v2-lp.md (hunk 4 above)
Comment (dev-1-lp-dollar) [reply UMIPs/umip-usd-uni-v2-lp.md:755]: okay

File: UMIPs/umip-usd-uni-v2-lp.md (hunk 5 above)
Comment (dev-1-lp-dollar) [reply UMIPs/umip-usd-uni-v2-lp.md:766]: My understanding is that using web3 or subgraph calls return the value at the end of the specific block. I.e. if i call `web3.queryContractState(199029)` it would return the state after block 199029 has been processed.

Because we are reading the reserves off chain at the end of the block (also applies if it queries at the beginning), then there is no potential for flashloan manipulation. FYI, I addressed a similar comment in [discourse here](https://discourse.umaproject.org/t/add-usd-uni-v2-wbtc-eth-as-a-price-identifier/148/9?u=dev-1-lpdollar).

There is the potential that a whale could suck it up and mess with the reserves by sending a huge buy or sell, but they would immediately be arbed by bots (costing them large amounts of money) and could not use a flash loan.

File: UMIPs/umip-usd-uni-v2-lp.md (hunk 7 above)
[the remark this answers is not in the document]
Comment (akhanubis) [reply UMIPs/umip-usd-uni-v2-lp.md:342]: I'm in favor of assuming 1 USDC = 1 USD.
What would be the price sources for USDC/USD?

File: UMIPs/umip-usd-uni-v2-lp.md (hunk 11 above)
[the remark this answers is not in the document]
Comment (akhanubis) [reply UMIPs/umip-usd-uni-v2-lp.md:134]: UMA team has released a template for price id UMIPs a few days ago: https://github.com/UMAprotocol/UMIPs/blob/master/price-identifier-template.md
One of the questions that the UMIP should provide an answer for is if you need an API key to query data from the diff price sources.

File: UMIPs/umip-usd-uni-v2-lp.md (hunk 7 above)
[the remark this answers is not in the document]
Comment (smb2796) [reply UMIPs/umip-usd-uni-v2-lp.md:342]: I agree. We've traditionally treated USDC as equivalent to USD, and it seems reasonable to do the same here.

File: UMIPs/umip-usd-uni-v2-lp.md (hunk 3 above)
Comment (Tommy1231232) [reply UMIPs/umip-usd-uni-v2-lp.md:22]: @smb2796 can we use the same price feed that was done for Badger?

File: UMIPs/umip-usd-uni-v2-lp.md (hunk 9 above)
[the remark this answers is not in the document]
Comment (Reinis-FRP) [reply UMIPs/umip-usd-uni-v2-lp.md:20]: Yes, it would be helpful to include brief explanation from the perspective of users (e.g. mint synth and sell for borrowers)

File: UMIPs/umip-usd-uni-v2-lp.md (hunk 12 above)
[the remark this answers is not in the document]
Comment (Reinis-FRP) [reply UMIPs/umip-usd-uni-v2-lp.md:139]: relative to the requested price feed timestamp seems reasonable, so better to add it

File: UMIPs/umip-usd-uni-v2-lp.md [hunk 14]
```diff
@@ -0,0 +1,766 @@
...
+The DVM currently does not support Uniswap V2 LP tokens.
+
+By enabling LP token price identifiers, UMA will open the door for the creation of synths with CFMM (constant function market maker) LP positions.
+
+The LP Dollar team will use the price identifiers to enable fixed borrowing costs for Uniswap V2 liquidity providers.
+
+## Data Sources & Price Feed Implementation
+
+| Uniswap V2 WBTC-ETH   |                                                                                                 |
```
Comment (smb2796) [inline UMIPs/umip-usd-uni-v2-lp.md:24]: The names of all of these price ids need to be changed to omit spaces and also include the currency that these are being paired with. Suggest: `USD/Uniswap_V2_WBTC_ETH`

File: UMIPs/umip-usd-uni-v2-lp.md [hunk 15]
```diff
@@ -0,0 +1,766 @@
+## Headers
+| UMIP-usd-uni-v2-lp-DRAFT    |                                                                                                                                          |
+|------------|------------------------------------------------------------------------------------------------------------------------------------------|
+| UMIP Title | Add USD UNI-V2 LP token price identifiers              |
```
Comment (smb2796) [inline UMIPs/umip-usd-uni-v2-lp.md:4]: Can you list all of the price identifier names here?

File: UMIPs/umip-usd-uni-v2-lp.md (hunk 3 above)
Comment (smb2796) [reply UMIPs/umip-usd-uni-v2-lp.md:22]: We should be able to yes. See this [pr](https://github.com/UMAprotocol/protocol/pull/2576/files) for an example of how to price LP tokens.

File: UMIPs/umip-usd-uni-v2-lp.md [hunk 16]
```diff
@@ -0,0 +1,766 @@
...
+     ```
+     
+     web3.js Response
+     ```
+     167105037364529719
+     ```
+
+3) Third, the ETHUSD price must be queried and used to calculate the USD value of the WETH reserves. 
+
+    The methodology to query the price of ETHUSD is similar to [UMIP-6](https://github.com/UMAprotocol/UMIPs/blob/master/UMIPs/umip-6.md), with a different exchange list and decimal precision.
```
Comment (smb2796) [inline UMIPs/umip-usd-uni-v2-lp.md:129]: Why is this change necessary?

File: UMIPs/umip-usd-uni-v2-lp.md [hunk 17]
```diff
@@ -0,0 +1,766 @@
...
+
+3) Third, the ETHUSD price must be queried and used to calculate the USD value of the WETH reserves. 
+
+    The methodology to query the price of ETHUSD is similar to [UMIP-6](https://github.com/UMAprotocol/UMIPs/blob/master/UMIPs/umip-6.md), with a different exchange list and decimal precision.  
+
+    Base Currency: ETH  
+    Quote Currency: USD  
+
+    Exchanges: Coinbase Pro (ETH:USD), Kraken (ETH:USD), Bitfinex (ETH:USD), Bitstamp (ETH:USD)  
+    Input Processing: None. Human intervention in extreme circumstances where the result differs from broad market consensus.
```
Comment (smb2796) [inline UMIPs/umip-usd-uni-v2-lp.md:135]: Please indent all of these

File: UMIPs/umip-usd-uni-v2-lp.md [hunk 18]
```diff
@@ -0,0 +1,766 @@
...
+    USD value of WETH reserves = WETH in reserves * WETH:USD = 
+    
+    = 97499.896966146357068372 * 1716.12 
+    
+    = 167321523.18154308
+    ```
+    
+4) Fourth, the BTC:USD price must be queried and used to calculate the USD value of the WBTC reserves. 
+
+    The methodology to query the price of BTC:USD is similar to [UMIP-7](https://github.com/UMAprotocol/UMIPs/blob/master/UMIPs/umip-7.md), with a different exchange list and decimal precision.
```
Comment (smb2796) [inline UMIPs/umip-usd-uni-v2-lp.md:155]: Why is this necessary? Also it would seem to make more sense to use the WBTCUSD price id, rather than BTCUSD

File: UMIPs/umip-usd-uni-v2-lp.md [hunk 19]
```diff
@@ -0,0 +1,766 @@
...
+    
+    = 3667.03647028 * 45938.30 = 168457421.48266372
+    ```
+
+5) Fifth, use the UNI-V2-WBTC-ETH total supply of LP tokens and USD value of reserves to calculate the USD value of each LP token (UNI-V2-WBTC-ETH:USD).
+
+    Example calculation of UNI-V2-WBTC-ETH:USD:
+  
+    ```
+    UNI-V2-WBTC-ETH:USD =
```
Comment (smb2796) [inline UMIPs/umip-usd-uni-v2-lp.md:182]: Great example

File: UMIPs/umip-usd-uni-v2-lp.md [hunk 20]
```diff
@@ -0,0 +1,766 @@
...
+    ```
+    USD:UNI-V2-WBTC-ETH = 
+  
+    (1 / UNI-V2-WBTC-ETH:USD) * 1e18
+  
+    = (1 / 2009388525.6835613) * 1e18 
+  
+    = 497663835.15
+  
+    = 497663835
```
Comment (smb2796) [inline UMIPs/umip-usd-uni-v2-lp.md:203]: This is confusing. The implementation should walk through how to calculate this price id before scaling. The voter dapp should abstract any necessary scaling away from a voter, and right now it seems like you are calculating the value of 1e18 USD in LP tokens.

File: UMIPs/umip-usd-uni-v2-lp.md [hunk 21]
```diff
@@ -0,0 +1,766 @@
...
+     ```
+
+     web3.js Response
+     ```
+     2572499047307646516
+     ```
+
+3) Third, the ETHUSD price must be queried and used to calculate the USD value of the WETH reserves. 
+
+    The methodology to query the price of ETHUSD is similar to [UMIP-6](https://github.com/UMAprotocol/UMIPs/blob/master/UMIPs/umip-6.md), with a different exchange list and decimal precision.
```
Comment (smb2796) [inline UMIPs/umip-usd-uni-v2-lp.md:311]: Since this is used in multiple places, I suggest defining a separate section for the alternative ETHUSD calculation to avoid redundancy.

File: UMIPs/umip-usd-uni-v2-lp.md (hunk 14 above)
Comment (smb2796) [reply UMIPs/umip-usd-uni-v2-lp.md:24]: I now see you added all of these at the end of the section, but please also add to `Summary`, `Headers` and the tope of each of these sections

File: UMIPs/umip-usd-uni-v2-lp.md [hunk 22]
```diff
@@ -0,0 +1,766 @@
...
+    
+    = 131583140.29680333
+    ```
+
+4) Fourth, the UNI:USD price must be queried and used to calculate the USD value of the UNI reserves.  
+
+    Base Currency: UNI  
+    Quote Currency: USD  
+
+    Exchanges: Coinbase Pro (UNI:USD), Kraken (UNI:USD), Bitfinex (UNI:USD)
```
Comment (smb2796) [inline UMIPs/umip-usd-uni-v2-lp.md:498]: Interesting that we have two UMIPs in the same week proposing different implementations for UNI/USD. Why did you choose these sources? The other price identifier uses Binance/OKEx/CB Pro, which seems more reasonable to me. Bitfinex and Kraken look to be lower volume than Binance and OKEx

File: UMIPs/umip-usd-uni-v2-lp.md (hunk 9 above)
[the remark this answers is not in the document]
Comment (dev-1-lp-dollar) [reply UMIPs/umip-usd-uni-v2-lp.md]: Added 👍

File: UMIPs/umip-usd-uni-v2-lp.md (hunk 12 above)
[the remark this answers is not in the document]
Comment (dev-1-lp-dollar) [reply UMIPs/umip-usd-uni-v2-lp.md]: Added 👍

File: UMIPs/umip-usd-uni-v2-lp.md (hunk 10 above)
[the remark this answers is not in the document]
Comment (dev-1-lp-dollar) [reply UMIPs/umip-usd-uni-v2-lp.md:21]: Added 👍

File: UMIPs/umip-usd-uni-v2-lp.md (hunk 22 above)
Comment (dev-1-lp-dollar) [reply UMIPs/umip-usd-uni-v2-lp.md:498]: Rationale was to use USD markets when possible rather than USDT markets

File: UMIPs/umip-usd-uni-v2-lp.md (hunk 18 above)
Comment (dev-1-lp-dollar) [reply UMIPs/umip-usd-uni-v2-lp.md:155]: Initially this was written because it seemed the USD:BTC price should carry more decimals than specified in UMIP-7. However, as the example calculation price shows, we just need BTC:USD, so it should be fine to reuse UMIP-7.

File: UMIPs/umip-usd-uni-v2-lp.md (hunk 22 above)
Comment (dev-1-lp-dollar) [reply UMIPs/umip-usd-uni-v2-lp.md:498]: i think swapping out Kraken for Binance is probably warranted given it only introduces one USDT market in the median calculation. And since Kraken has the lowest volume of the three.

File: UMIPs/umip-usd-uni-v2-lp.md (hunk 16 above)
Comment (dev-1-lp-dollar) [reply UMIPs/umip-usd-uni-v2-lp.md]: Changed to use UMIP-6

File: UMIPs/umip-usd-uni-v2-lp.md (hunk 18 above)
Comment (dev-1-lp-dollar) [reply UMIPs/umip-usd-uni-v2-lp.md]: Changed to use UMIP-7

File: UMIPs/umip-usd-uni-v2-lp.md (hunk 13 above)
[the remark this answers is not in the document]
Comment (dev-1-lp-dollar) [reply UMIPs/umip-usd-uni-v2-lp.md]: Added language to be more specific about decimals in calculation steps, and per other comments, have switched to using UMIP-7 for BTC:USD

File: UMIPs/umip-usd-uni-v2-lp.md (hunk 20 above)
Comment (dev-1-lp-dollar) [reply UMIPs/umip-usd-uni-v2-lp.md]: Edited the example to take this into account

File: UMIPs/umip-usd-uni-v2-lp.md (hunk 18 above)
Comment (dev-1-lp-dollar) [reply UMIPs/umip-usd-uni-v2-lp.md]: Regarding WBTCUSD price id, this UMIP seems to be missing clarification on the ETH:USD conversation (unless I missed it), and seems to add a decent amount of complexity. Unless there are objections, we believe BTC:USD will serve users (and liquidatiors and disputers) of the price id well.

File: UMIPs/umip-usd-uni-v2-lp.md (hunk 21 above)
Comment (dev-1-lp-dollar) [reply UMIPs/umip-usd-uni-v2-lp.md]: A lot of the text here was taken out with the reference to UMIP-6. I think it is fine to leave this as the redundancy helps the examples flow better (i.e. rather than jumping from place to place in the doc).

File: UMIPs/umip-usd-uni-v2-lp.md (hunk 22 above)
Comment (dev-1-lp-dollar) [reply UMIPs/umip-usd-uni-v2-lp.md]: Added Binance, took out Kraken

File: UMIPs/umip-usd-uni-v2-lp.md (hunk 14 above)
Comment (smb2796) [reply UMIPs/umip-usd-uni-v2-lp.md:24]: @dev-1-lp-dollar thoughts on this name change?

File: UMIPs/umip-usd-uni-v2-lp.md (hunk 3 above)
Comment (smb2796) [reply UMIPs/umip-usd-uni-v2-lp.md:22]: To @Reinis-FRP's point. Due to the amount of price ids in this umip, I think it is fine to combine `Markets & Data Sources` with the voting `Implementation`. I do think a separate `Price Feed Implementation` would work well however. In the `Price Feed Implementation` section, you can just explain that it would be the same model for all of these price identifiers. These should use a combination of the [CryptowatchPriceFeed](https://github.com/UMAprotocol/protocol/blob/master/packages/financial-templates-lib/src/price-feed/CryptoWatchPriceFeed.js), [LPPriceFeed](https://github.com/UMAprotocol/protocol/blob/master/packages/financial-templates-lib/src/price-feed/LPPriceFeed.js) and [ExpressionPriceFeed](https://github.com/UMAprotocol/protocol/blob/master/packages/financial-templates-lib/src/price-feed/ExpressionPriceFeed.js).

As a meta point, I think it is becoming apparent that we should make it a requirement that new price identifiers need a [default price feed config](https://github.com/UMAprotocol/protocol/blob/master/packages/financial-templates-lib/src/price-feed/DefaultPriceFeedConfigs.js) even when using existing price feeds. @dev-1-lp-dollar if you are able to submit a pr for this and link it in `Price Feed Implementation` that would be great. If not, I think that is fine given that this hasn't historically been a requirement.

File: UMIPs/umip-usd-uni-v2-lp.md (hunk 20 above)
Comment (smb2796) [reply UMIPs/umip-usd-uni-v2-lp.md]: Thanks, looks good to me.

File: UMIPs/umip-usd-uni-v2-lp.md (hunk 21 above)
Comment (smb2796) [reply UMIPs/umip-usd-uni-v2-lp.md]: That's fine

File: UMIPs/umip-usd-uni-v2-lp.md (hunk 22 above)
Comment (smb2796) [reply UMIPs/umip-usd-uni-v2-lp.md]: That seems reasonable to me

File: UMIPs/umip-usd-uni-v2-lp.md (hunk 18 above)
Comment (smb2796) [reply UMIPs/umip-usd-uni-v2-lp.md]: It has [ETHUSD](https://github.com/UMAprotocol/UMIPs/blob/master/UMIPs/umip-51.md#implementation) conversion. I'd argue that given the custodial nature of WBTC and given its tendency to trade off peg to BTC, it is necessary to calculate with WBTCUSD rather than BTCUSD. You could always define your own WBTCUSD implementation here if umip-51 does not seem suitable. What do you think?

**Comment by smb2796** [comment]
@dev-1-lp-dollar @dev-2-lp-dollar curious to hear your thoughts on the comment I left around WBTC. Other than that, I think the main outstanding issue is the price identifier [names](https://github.com/UMAprotocol/UMIPs/pull/170#discussion_r580274794).

File: UMIPs/umip-usd-uni-v2-lp.md [hunk 23]
```diff
@@ -0,0 +1,700 @@
+## Headers
+| UMIP-usd-uni-v2-lp-DRAFT    |                                                                                                                                          |
+|------------|------------------------------------------------------------------------------------------------------------------------------------------|
+| UMIP Title | Add USD UNI-V2 WBTC-ETH, USDC-ETH, UNI-ETH, UMA-ETH LP token price identifiers              |
```
Comment (smb2796) [inline UMIPs/umip-usd-uni-v2-lp.md:4]: Can you please update these names to match the ones in `Technical Specifications`

File: UMIPs/umip-usd-uni-v2-lp.md [hunk 24]
```diff
@@ -0,0 +1,700 @@
...
+    = (1 / 520.3440264478902)
+  
+    0.001921805477092653  
+    ```
+
+
+## Technical Specifications
+
+Price Identifier Name: USD/UNI-V2-WBTC-ETH  
+Base Currency: USD
```
Comment (smb2796) [inline UMIPs/umip-usd-uni-v2-lp.md:662]: Please bullet all of these

File: UMIPs/umip-usd-uni-v2-lp.md [hunk 25]
```diff
@@ -0,0 +1,700 @@
...
+    USD value of WETH reserves = WETH in reserves * WETH:USD = 
+    
+    = 1350.358508316793260065 * 1716.12 
+    
+    = 2317377.24329262
+    ```
+
+4) Fourth, the UMA:USD price must be queried and used to calculate the USD value of the UMA reserves.  
+
+    Base Currency: UMA
```
Comment (smb2796) [inline UMIPs/umip-usd-uni-v2-lp.md:609]: Please bullet these and all similar sections

File: UMIPs/umip-usd-uni-v2-lp.md (hunk 18 above)
Comment (dev-1-lp-dollar) [reply UMIPs/umip-usd-uni-v2-lp.md]: I haven't seen data to suggest WBTC trades off peg by more than minting or redemption fees, but would be interested to see it. Can you share this? 

I have two thoughts / concerns that makes me hesitant: 

1) As a minter, I would anticipate 1 WBTC to be equal to 1 BTC.
2) UMIP-51 looks robust enough for a price oracle but seems to add non-trivial overhead to calculating the liquidation price.

So, I suppose the amount by which it trades off peg would determine the robustness here. I wouldn't want users to get liquidated because it traded off peg and BitGo wasn't hacked and WBTC was redeemable for 1 BTC - redemption cost. But I also see the value in handling the custodial nature here.

Comment (smb2796) [reply UMIPs/umip-usd-uni-v2-lp.md]: Fair points. I generally meant the difference caused by minting + redemption fees, but you're right this difference is somewhat negligible.

I'd generally argue that the actual act of calculating the WBTCUSD price is no more difficult than for BTCUSD, but I recognize that this is subjective.

If some WBTC black swan event did happen, WBTC started trading way off peg from BTC and this price id was not representative of the actual LP token values, voters could always come to a consensus around using the WBTC rates instead. This level of flexibility is one of the coolest things about the DVM imo, and gives us an option if this ever does become a bigger concern.

Overall you've convinced me and I feel fine with leaving it as is!

**Review by smb2796 — APPROVED** [review]
Looks like all comments have been addressed except for a few small formatting issues. Will fix those on my end. Going to merge in and move to `Last Call`.
